### DISCUSSION TITLE
CLI: -opaque isn't the opposite of -transparent is something else able to do that?

### DISCUSSION THREAD

**Discussion by tatarize** [ImageMagick/ImageMagick#3298] [Development]
`magick convert input.png -fuzz 40% -transparent blue output.png`

Then:

`magick convert output.png -fuzz 40% -opaque blue output2.png`

Does not recreate the image. `"The -transparent operator is exactly the same as -opaque but replaces the matching color with transparency rather than the current -fill color setting."` but what I want to do is basically make the fuzzy-matched pixels transparent. Then equally find the purely transparent fuzzy-matched pixels and make them opaque again.

I can do:
`magick convert output.png -alpha off -alpha on output2.png`

And I'll get my original pixels back, but is there something else that actually *is* the opposite command as -transparent. Find the pixels that match this color set their alpha to 0% and find these pixels that match this color (ignoring that they are transparent) and set their alpha to 100%?

**Comment by snibgo** [comment]
For IM v7, I suggest you use "magick", not "magick convert".

"-transparent" and "-opaque" are not opposites of each other. They do different things.

"-transparent" makes opaque pixels transparent (without changing colours). It is not reversible, because the image might already contain transparent pixels of the specified colour.

"-opaque" is used with a "-fill", where the default fill colour is white. It changes opaque pixels to the fill colour. Again, it is not reversible, because the image may already contain pixels of the fill colour. If "-fuzz" is used, all pixels that were within the fuzz distance will change to be the same colour, so this also makes reversing impossible.

I'm not sure what you want. I think you want to make some opaque pixels transparent, then make those same pixels opaque. So there is no overall change, no effect at all.

**Reply by tatarize** [reply]
Yes. really I want something that actually *is* the opposite of -transparent. Find the pixels even if they are transparent and still that color, and set their alpha values to 100%. 

In reality this would be undoing some changes. I could reset the alpha after several color removes rather than have a restore all the pixels. But, I only kinda wanted to restore some colors. If nothing else -opaque should pretend to be the inverse in the docs since they *are* doing pretty radically different things.

**Comment by snibgo** [comment]
> Yes. really I want something that actually is the opposite of -transparent.

As I have explained, nothing is the opposite of "-transparent" because "-transparent" is not reversible.

Perhaps you want to make some pixels transparent, then do some other operation, then make those same pixels opaque. If so, then you need to record which pixels have been made transparent. You could do that with a mask, and eventually use that mask to make pixels opaque.

If your "other operation" does not change the alpha channel, then you could simply save the alpha channel and restore it afterwards. For example, Windows BAT syntax:
```
magick ^
  in.png ^
  ( +clone -alpha extract -write mpr:ALPH +delete ) ^
  -fuzz 40%% -transparent blue ^
  {other_operations} ^
  mpr:ALPH -alpha off -compose CopyOpacity -composite ^
  out.png
```

**Reply by tatarize** [reply]
You could almost use that to to make a reverse of transparent. Which actually is reversible. If you do `-alpha off -alpha on` you get all your original pixels back before you did the -transparent. The only destruction it's done is alpha manipulation.

Find pixels.
Set alpha 0.

I want:

Find pixels.
Set alpha 255.

You can restore all the original pixels by throwing out the alpha. Then use the purely opaque version to find those values again, then set their alpha. I just want the pixels matching a given color to be restored rather than all of them.

**Reply by tatarize** [reply]
Now, it is possible to do:

Save a copy.
Restore all original pixels.
Find pixels.
Mask the transparent mask to set the alpha back in there.

But, the pixel data is still there. And you could find what pixels a transparent operation would find, then use that to create a mask and apply that back to the original copy of the original image and use that to unset the mask without having initially solved it.
